File: go/radixsort/radixsort.go
package main

import (
	"fmt"
	"math"
	"strconv"
)


// Time -> 0(d * (b+n)) Space -> 0(b+n) where d is the max len of digit and b is the base of the nums(base 10)
func RadixSort(array []int) []int {
	digit, maxDigit := 0, Max(array)
	for digit < maxDigit {
		CountSort(array, digit)
		digit++
	}
	return array
}

func CountSort(array []int, digit int) {
	count := make([]int, 10)
	newarray := make([]int, len(array))
	digitCol := int(math.Pow(10, float64(digit)))
	
	for _, v := range array {
		idx := (v / digitCol) % 10
		count[idx] += 1
	}

	for idx:=1; idx<len(count); idx++ {
		count[idx] += count[idx-1]
	}

	for idx:=len(array)-1; idx>=0; idx-- {
		countIdx := (array[idx] / digitCol) % 10
		count[countIdx] -= 1
		newarray[count[countIdx]] = array[idx]
	}
	copy(array, newarray)
}

func Max(array []int) int {
	max := math.MinInt
	for _, v := range array {
		if v > max{max = v}
	}
	return len(strconv.Itoa(max))
}

func main() {
	fmt.Println(RadixSort([]int{234, 12, 90, 5, 7, 22, 52, 202, 75}))
}

File: python/radixsort.py
# Time -> 0(d*(b+n)) Space -> 0(b+n) where d is the max num of digit and b is the base of the nums(base 10)
def radixsort(array):
    maxDig = max(array)
    digit, maxDig = 0, len(str(maxDig))
    while digit < maxDig:
        countsort(array, digit)
        digit += 1
        
    return array

def countsort(array, digit):
    count = [0] * 10
    newArray = [0] * len(array)
    digitCol = 10 ** digit
    
    for num in array:
        idx = (num // digitCol) % 10
        count[idx] += 1
        
    for i in range(1, len(count)):
        count[i] += count[i-1]
    
    for i in reversed(range(len(array))):
        countIdx = (array[i] // digitCol) % 10
        count[countIdx] -= 1
        newArray[count[countIdx]] = array[i]
        
    array[:] = newArray
    
    
print(radixsort([234, 12, 90, 5, 7, 22, 52, 202, 75]))
